File: core/src/main/java/javabot/dao/ApiDao.java
package javabot.dao;

import java.util.List;
import javax.inject.Inject;

import javabot.javadoc.JavadocApi;
import javabot.criteria.JavadocApiCriteria;
import org.bson.types.ObjectId;

public class ApiDao extends BaseDao<JavadocApi> {
  @Inject
  public JavadocClassDao classDao;

  protected ApiDao() {
    super(JavadocApi.class);
  }

  public JavadocApi find(final String name) {
    JavadocApiCriteria criteria = new JavadocApiCriteria(ds);
    criteria.upperName().equal(name.toUpperCase());
    return criteria.query().get();
  }

  public void delete(final JavadocApi api) {
    classDao.deleteFor(api);
    super.delete(api);
  }

  public void delete(final ObjectId id) {
    delete(find(id));
  }

  @Override
  public List<JavadocApi> findAll() {
    return ds.createQuery(JavadocApi.class).order("name").asList();
  }
}

File: src/main/kotlin/javabot/dao/ApiDao.kt
package javabot.dao

import dev.morphia.Datastore
import dev.morphia.DeleteOptions
import dev.morphia.query.FindOptions
import dev.morphia.query.Sort
import dev.morphia.query.filters.Filters.eq
import java.util.Locale
import javabot.model.javadoc.JavadocApi
import javabot.model.javadoc.JavadocClass
import javabot.model.javadoc.JavadocField
import javabot.model.javadoc.JavadocMethod
import javax.inject.Inject
import org.bson.types.ObjectId
import org.slf4j.LoggerFactory

class ApiDao @Inject constructor(ds: Datastore) : BaseDao<JavadocApi>(ds, JavadocApi::class.java) {
    companion object {
        private val LOG = LoggerFactory.getLogger(ApiDao::class.java)
    }

    fun find(name: String) =
        ds.find(JavadocApi::class.java)
            .filter(eq("upperName", name.uppercase(Locale.getDefault())))
            .first()

    fun delete(api: String) {
        find(api)?.let { delete(it) }
    }

    override fun delete(id: ObjectId?) {
        find(id)?.let { delete(it) }
    }

    fun delete(api: JavadocApi) {
        LOG.debug("Dropping fields from " + api.name)

        ds.find(JavadocField::class.java)
            .filter(eq("apiId", api.id))
            .delete(DeleteOptions().multi(true))

        LOG.debug("Dropping methods from " + api.name)
        ds.find(JavadocMethod::class.java)
            .filter(eq("apiId", api.id))
            .delete(DeleteOptions().multi(true))

        LOG.debug("Dropping classes from " + api.name)
        ds.find(JavadocClass::class.java)
            .filter(eq("apiId", api.id))
            .delete(DeleteOptions().multi(true))

        super.delete(api)
    }

    override fun findAll(): List<JavadocApi> {
        return ds.find(JavadocApi::class.java)
            .iterator(FindOptions().sort(Sort.ascending("name")))
            .toList()
    }
}
